File: backend/app/services/deep_research/prompts.py
"""LLM prompt templates for Deep Research analysis pipeline.

All prompts use XML-tag delimiters for paper content to prevent
prompt injection (research pitfall #5). Templates use double-brace
escaping for JSON output structure in f-strings.

Reference: AI-Scientist generate_ideas.py, gpt-researcher analysis prompts.
"""

# ─── Abstract-Level Screening (Haiku) ─────────────────────────────────────


def build_tldr_prompt(
    title: str,
    abstract: str,
    year: int | None,
    venue: str | None,
) -> list[dict]:
    """Build TLDR screening prompt for abstract-level analysis.

    Returns bilingual summary, methods, datasets, metrics, paper type.
    Truncates abstract to 1000 chars for cost control.
    """
    safe_abstract = (abstract or "")[:1000]
    safe_year = str(year) if year else "unknown"
    safe_venue = venue or "unknown"

    content = f"""Analyze the following academic paper and provide a structured analysis.

<paper>
<title>{title}</title>
<abstract>{safe_abstract}</abstract>
<year>{safe_year}</year>
<venue>{safe_venue}</venue>
</paper>

Return a JSON object with exactly these fields:
{{
  "tldr_en": "One-sentence summary in English (max 100 words)",
  "tldr_zh": "One-sentence summary in Chinese (max 100 characters)",
  "methods": ["method1", "method2"],
  "datasets": ["dataset1", "dataset2"],
  "key_metrics": {{"metric_name": "value"}},
  "paper_type": "empirical|theoretical|survey|application|methodology"
}}

Return ONLY valid JSON, no additional text."""

    return [{"role": "user", "content": content}]


# ─── Full-Text Deep Analysis (Sonnet) ─────────────────────────────────────


def build_deep_analysis_prompt(
    title: str,
    abstract: str | None,
    parsed_content: str | None,
) -> list[dict]:
    """Build deep analysis prompt for top-N papers with full text.

    Extracts detailed methodology, contributions, limitations, future work.
    Uses parsed PDF content when available, falls back to abstract.
    """
    text_content = parsed_content or abstract or "No content available"
    # Truncate to ~4000 chars to fit context and control costs
    safe_content = text_content[:4000]

    content = f"""Perform a detailed analysis of this academic paper.

<paper>
<title>{title}</title>
<content>{safe_content}</content>
</paper>

Return a JSON object with exactly these fields:
{{
  "detailed_methodology": "Detailed description of the methodology used (2-3 paragraphs)",
  "key_contributions": ["contribution 1", "contribution 2", "contribution 3"],
  "limitations": ["limitation 1", "limitation 2"],
  "future_work": ["direction 1", "direction 2"]
}}

Return ONLY valid JSON, no additional text."""

    return [{"role": "user", "content": content}]


# ─── Relationship Classification (Haiku) ──────────────────────────────────


def build_relationship_prompt(
    title_a: str,
    abstract_a: str,
    title_b: str,
    abstract_b: str,
) -> list[dict]:
    """Build pairwise relationship classification prompt.

    Classifies how paper_b relates to paper_a. Only called for
    citation-connected pairs (not O(n^2)).
    Truncates abstracts to 500 chars each.
    """
    safe_abstract_a = (abstract_a or "")[:500]
    safe_abstract_b = (abstract_b or "")[:500]

    content = f"""Given two academic papers, classify their relationship.

<paper_a>
<title>{title_a}</title>
<abstract>{safe_abstract_a}</abstract>
</paper_a>

<paper_b>
<title>{title_b}</title>
<abstract>{safe_abstract_b}</abstract>
</paper_b>

Classify the relationship of paper_b relative to paper_a as exactly ONE of:
- "improvement": paper_b improves upon paper_a's method
- "comparison": paper_b compares against paper_a
- "survey": paper_b surveys/reviews paper_a's area
- "application": paper_b applies paper_a's method to a new domain
- "theoretical_basis": paper_a provides theoretical foundation for paper_b
- "unrelated": no meaningful relationship

Return JSON: {{"relationship": "<type>", "confidence": 0.0, "explanation": "brief reason"}}

Return ONLY valid JSON, no additional text."""

    return [{"role": "user", "content": content}]

File: project-7411174/src/lib/deep-research/prompts.ts
/**
 * LLM prompt templates for client-side deep research.
 * Ported from backend prompts.py with XML-tag delimiters.
 */

import type { AnalyzedPaper, ScoredPaper } from "./types";

/**
 * Build TLDR screening prompt for a batch of papers.
 * Processes multiple papers in one LLM call for efficiency.
 */
export function buildBatchAnalysisPrompt(papers: ScoredPaper[]): string {
  const paperEntries = papers.map((p, i) => {
    const safeAbstract = (p.abstract || "无摘要")
      .slice(0, 600)
      .replace(/[<>]/g, "");
    return `<paper index="${i + 1}">
<title>${p.title.replace(/[<>]/g, "")}</title>
<abstract>${safeAbstract}</abstract>
<year>${p.year ?? "unknown"}</year>
<venue>${p.venue || "unknown"}</venue>
</paper>`;
  }).join("\n\n");

  return `Analyze the following ${papers.length} academic papers. For each paper, provide a structured analysis.

${paperEntries}

Return a JSON array with one object per paper (in order), each containing:
{
  "index": 1,
  "tldr_en": "One-sentence summary in English (max 80 words)",
  "tldr_zh": "一句话中文总结（最多80字）",
  "methods": ["method1", "method2"],
  "datasets": ["dataset1", "dataset2"],
  "paper_type": "empirical|theoretical|survey|application|methodology"
}

Return ONLY a valid JSON array, no additional text.`;
}

/**
 * Build relationship classification prompt for paper pairs.
 */
export function buildRelationshipPrompt(
  pairs: { a: AnalyzedPaper; b: AnalyzedPaper }[],
): string {
  const entries = pairs.map((pair, i) => {
    const absA = (pair.a.abstract || "").slice(0, 300).replace(/[<>]/g, "");
    const absB = (pair.b.abstract || "").slice(0, 300).replace(/[<>]/g, "");
    return `<pair index="${i + 1}">
<paper_a>${pair.a.title.replace(/[<>]/g, "")}: ${absA}</paper_a>
<paper_b>${pair.b.title.replace(/[<>]/g, "")}: ${absB}</paper_b>
</pair>`;
  }).join("\n\n");

  return `Classify the relationship between each pair of academic papers.

${entries}

For each pair, classify paper_b's relationship to paper_a as ONE of:
- "improvement": paper_b improves upon paper_a's method
- "comparison": paper_b compares against paper_a
- "survey": paper_b surveys paper_a's area
- "application": paper_b applies paper_a's method to a new domain
- "theoretical_basis": paper_a provides theoretical foundation for paper_b
- "unrelated": no meaningful relationship

Return a JSON array:
[{"index": 1, "relationship": "<type>", "confidence": 0.8, "explanation": "brief reason"}]

Return ONLY valid JSON array.`;
}

/**
 * Build gap detection prompt from analyzed corpus.
 */
export function buildGapDetectionPrompt(
  direction: string,
  papers: AnalyzedPaper[],
): string {
  // Build corpus summary
  const methodCounts: Record<string, number> = {};
  const yearCounts: Record<number, number> = {};

  for (const p of papers) {
    for (const m of p.methods) {
      methodCounts[m] = (methodCounts[m] || 0) + 1;
    }
    if (p.year) {
      yearCounts[p.year] = (yearCounts[p.year] || 0) + 1;
    }
  }

  const topMethods = Object.entries(methodCounts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 20)
    .map(([m, c]) => `${m}: ${c}`)
    .join("\n");

  const paperTypes = papers.reduce(
    (acc, p) => {
      acc[p.paperType] = (acc[p.paperType] || 0) + 1;
      return acc;
    },
    {} as Record<string, number>,
  );

  const summaryLines = [
    `Total papers: ${papers.length}`,
    `Year range: ${Math.min(...papers.filter(p => p.year).map(p => p.year!))} - ${Math.max(...papers.filter(p => p.year).map(p => p.year!))}`,
    `Paper types: ${Object.entries(paperTypes).map(([t, c]) => `${t}(${c})`).join(", ")}`,
    `Top TLDRs (sample):`,
    ...papers.slice(0, 5).map(p => `  - ${p.tldrEn}`),
  ].join("\n");

  return `You are analyzing a corpus of ${papers.length} academic papers in the area of "${direction}".

<corpus_summary>
${summaryLines}
</corpus_summary>

<method_frequencies>
${topMethods}
</method_frequencies>

Identify as many as possible (aim for 10+ gaps):
1. Research gaps: areas mentioned but not deeply explored (at least 5)
2. Underexplored combinations: methods/techniques that haven't been combined yet (at least 3)
3. Missing evaluations: datasets or metrics not yet applied to promising methods (at least 3)
4. Methodological gaps: limitations in current approaches that need addressing
5. Application gaps: domains where existing methods haven't been applied

Be thorough and specific. Each gap should be actionable — a researcher should be able to start a project based on it.

Return JSON:
{
  "gaps": [{"description": "...", "evidence": "...", "potential_impact": "high|medium|low"}],
  "underexplored": [{"combination": "...", "why_promising": "..."}],
  "missing_evaluations": [{"method": "...", "missing": "..."}]
}

Return ONLY valid JSON.`;
}

/**
 * Build report generation prompt.
 */
export function buildReportPrompt(
  direction: string,
  papers: AnalyzedPaper[],
  gapAnalysis: { gaps: { description: string }[]; underexplored: { combination: string }[] } | null,
): string {
  // Build numbered source list
  const numberedSources = papers
    .slice(0, 30) // Limit to top 30 for context window
    .map((p, i) => {
      const authors = p.authors.slice(0, 3).join(", ");
      return `[${i + 1}] ${authors} (${p.year ?? "n.d."}). "${p.title}". ${p.venue || ""}. ${p.tldrEn || ""}`;
    })
    .join("\n");

  const gapsSummary = gapAnalysis
    ? `\n已识别的研究空白:\n${gapAnalysis.gaps.map(g => `- ${g.description}`).join("\n")}\n已识别的未探索组合:\n${gapAnalysis.underexplored.map(u => `- ${u.combination}`).join("\n")}`
    : "";

  return `Write a comprehensive literature review on "${direction}" in Chinese (中文).

The review should include these sections:
1. 概述 (Overview) — 2-3 paragraphs introducing the field
2. 方法论综述 (Methodology Review) — main approaches and their evolution
3. 关键发现 (Key Findings) — important results and contributions
4. 研究空白与未来方向 (Gaps and Future Directions) — what's missing and what's next

${gapsSummary}

Sources (use [1], [2], etc. for inline citations):
${numberedSources}

Rules:
- Write in Chinese (中文)
- Use [1], [2], ..., [n] inline to cite sources
- Every factual claim MUST have at least one citation
- Write analytical prose, NOT bullet points
- Each section should be 2-4 paragraphs
- Use ## for section headings
- Do NOT include a separate References section (it will be added automatically)

Return the complete review text with inline citations.`;
}
